### PR TITLE
Fixed bibliography latex field

### PR DESCRIPTION
**Pull request by devhassanmahmood**

### CODE REVIEW

File: app/controllers/proposals_controller.rb [hunk 1]
```diff
@@ -73,8 +74,9 @@ def latex_field
 
     @proposal = Proposal.find_by(id: prop_id)
     @year = @proposal&.year || Date.current.year.to_i + 2
+    check_file
     @latex_infile = File.read("#{Rails.root}/tmp/#{latex_temp_file}")
-    @latex_infile = LatexToPdf.escape_latex(@latex_infile) if @proposal.no_latex
+    @latex_infile = LatexToPdf.escape_latex(@latex_infile) unless @proposal.no_latex
```
Comment (brentkearney) [inline app/controllers/proposals_controller.rb:79]: If no_latex is selected, then we want to escape the latex

File: app/controllers/proposals_controller.rb [hunk 2]
```diff
@@ -170,4 +172,17 @@ def set_careers
     @careers = Person.where(id: @proposal.participants.pluck(:person_id))
                      .pluck(:academic_status)
   end
+
+  def check_file
```
Comment (brentkearney) [inline app/controllers/proposals_controller.rb:176]: create_latex_file would be a more suitable name

File: app/controllers/submit_proposals_controller.rb [hunk 3]
```diff
@@ -114,4 +115,15 @@ def update_proposal_ams_subject_code
...
+    temp_file = "propfile-#{current_user.id}-#{@proposal.id}.tex"
+    return if File.exist?("#{Rails.root}/tmp/#{temp_file}")
+
+    @latex_infile = ProposalPdfService.new(@proposal.id, temp_file, 'all')
+                                      .generate_latex_file.to_s
+    File.new("#{Rails.root}/tmp/#{temp_file}", 'w') do |io|
+      io.write(@latex_infile)
+    end
+  end
 end
```
Comment (brentkearney) [inline app/controllers/submit_proposals_controller.rb:129]: Duplicate code should be extracted to the PdfService class

File: app/controllers/submitted_proposals_controller.rb [hunk 4]
```diff
@@ -115,6 +115,7 @@ def create_pdf_file
                                 inline: prop_latex.to_s, formats: [:pdf]
 
     @pdf_path = "#{Rails.root}/tmp/submit-#{DateTime.now.to_i}.pdf"
+    check_file
     File.open(@pdf_path, "w:UTF-8") do |file|
```
Comment (brentkearney) [inline app/controllers/submitted_proposals_controller.rb:119]: If check_file returns nil, @pdf_path will be undefined...

File: app/services/proposal_pdf_service.rb [hunk 5]
```diff
@@ -63,7 +68,7 @@ def all_proposal_fields
     proposal_organizers
     proposal_locations
     proposal_subjects
-    proposal_bibliography
+    proposal_bibliography unless proposal.no_latex
     user_defined_fields
```
Comment (brentkearney) [inline app/services/proposal_pdf_service.rb:72]: bibliography should appear at the bottom of the document, so below user_defined_fields

**Review by brentkearney — CHANGES_REQUESTED** [review]

File: app/controllers/proposals_controller.rb [hunk 6]
```diff
@@ -170,4 +172,17 @@ def set_careers
...
+    return if File.exist?("#{Rails.root}/tmp/#{latex_temp_file}")
+
+    if params[:action] == "latex_field"
+      input = latex_params[:latex]
+      @data = ProposalPdfService.new(@proposal.id, latex_temp_file, input)
+                                .generate_latex_file
+    end
+    File.new("#{Rails.root}/tmp/#{latex_temp_file}", 'w') do |io|
+      io.write(@data)
+    end
```
Comment (brentkearney) [inline app/controllers/proposals_controller.rb:186]: `ProposalPdfService. generate_latex_file` already writes latex_temp_file to file

File: app/services/proposal_pdf_service.rb [hunk 7]
```diff
@@ -8,22 +8,27 @@ def initialize(proposal_id, file, input)
...
 
   def to_s
+    unless File.exist?("#{Rails.root}/tmp/#{@temp_file}")
+      File.new("#{Rails.root}/tmp/#{@temp_file}", 'w') do |io|
+        io.write(@input)
+      end
+    end
     latex_infile = File.read("#{Rails.root}/tmp/#{@temp_file}")
-    latex_infile = LatexToPdf.escape_latex(latex_infile) if @proposal.no_latex
+    latex_infile = LatexToPdf.escape_latex(latex_infile) unless @proposal.no_latex
```
Comment (brentkearney) [inline app/services/proposal_pdf_service.rb:31]: text should be escaped IF no_latex is true

File: app/controllers/proposals_controller.rb [hunk 8]
```diff
@@ -47,20 +47,21 @@ def latex_input
...
     head :ok
   end
 
   # GET /proposals/:id/rendered_proposal.pdf
   def latex_output
     proposal_id = params[:id]
-    ProposalPdfService.new(proposal_id, latex_temp_file, 'all')
-                      .generate_latex_file
+    @data = ProposalPdfService.new(proposal_id, latex_temp_file, 'all')
+                              .generate_latex_file
```
Comment (brentkearney) [inline app/controllers/proposals_controller.rb:60]: @data is an instance of ProposalPdfService. I don't see where @data is used...

File: app/services/proposal_pdf_service.rb [hunk 9]
```diff
@@ -116,8 +121,10 @@ def proposal_subjects
   end
 
   def proposal_bibliography
-    @text << "\\subsection*{Bibliography}\n\n"
-    @text << "\\noindent #{LatexToPdf.escape_latex(proposal.bibliography)}\n\n"
+    if proposal.bibliography.present?
+      @text << "\\subsection*{Bibliography}\n\n"
+      @text << "\\noindent #{LatexToPdf.escape_latex(proposal.bibliography)}\n\n"
```
Comment (brentkearney) [inline app/services/proposal_pdf_service.rb:126]: Bibliography will always be LaTeX, so should not be escaped

**Review by birs-math — APPROVED** [review]
Regression bug: clicking the No LaTex checkbox no longer saves the proposal, so "Save as Draft" must be clicked

**Pull request merged**